corehq/apps/users/tests/test_email_events.py

```diff
@@ -36,3 +42,51 @@ def test_bounced_event(self):
         invite = Invitation.objects.get(uuid=self.invite.uuid)
 
         self.assertEqual(invite.email_status, InvitationStatus.BOUNCED)
+
+
+class TestUserLimitWarnings(TestSnsEmailBase):
+    def setUp(self):
+        self.domain = create_domain("testapp")
+        initialize_domain_with_default_roles(self.domain.name)
+        self.recipients = ["orange@gmail.com", "purple@gmail.com"]
+        self.domain_admin = WebUser.create(self.domain.name, self.recipients[0], "123", None, None, is_admin=True)
+        self.domain_admin.save()
+        self.billing_admin_user = WebUser.create(self.domain.name, self.recipients[1], "123", None, None)
+        role = UserRole.objects.get(domain=self.domain, name=UserRolePresets.BILLING_ADMIN)
+        self.billing_admin_user.set_role(self.domain.name, role.get_qualified_id())
+        self.billing_admin_user.save()
+
+    def tearDown(self):
+        self.domain_admin.delete(self.domain.name, None)
+        self.billing_admin_user.delete(self.domain.name, None)
+        self.domain.delete()
+
+    def test_limit_emails(self):
```

Comment:
I'm tempted to roll the last two into the respective cases that also test whether they send at the right triggers (at/past threshold and at/past the maximum) because if they were their own tests, I'd have to do that again + make sure that it doesn't send while remaining above the 90% or 100% threshold.  Would that be an issue of cramming too many things into a single test again?

Reply:
Also, in general, how granular should these tests be? I haven't been great about adding tests with new features so I'm trying to get a feel for how much is enough, going forward.

Reply:
Ideally, every condition is its own test case. Even if things are related, it doesn't mean they aren't implemented in a way that an error in one part of the condition implies success or failure in the other part of the condition. Putting them in the same test case falls back into the same problems with stopping at the first failure and inheriting state. If the code strongly makes you want to avoid re-creating data for separate test cases, that can be a good indication that refactoring is necessary to make the code easier to test.

As a shortcut, you might hear someone say that a test should only have one assertion. That is incorrect -- a test should only be testing one thing, and sometimes that one thing requires multiple assertions (i.e that all items in an array match).  A useful shortcut, however, is seeing multiple assertions _with test setup between them_. That is two tests masquerading as one. 

If you're not familiar with it, check out the [AAA (Arrange, Act, Assert) pattern](https://robertmarshall.dev/blog/arrange-act-and-assert-pattern-the-three-as-of-unit-testing/).

I might be missing something, but in this case, I actually think the code seems easy enough to test? Because your function takes in all the relevant variables (`plan_limit`, `user_count`, and `prev_count`), you don't need to trigger a previous email before you can verify that a second call will skip the email -- you just have to make sure `prev_count` is above the threshold you're testing.